File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool dfs(int r, int c, int i, string& word, vector<vector<char>>& board, vector<pair<int, int>>& dir){
        int n = board.size();
        int m = board[0].size();
        int k = word.size();

        if(i == k) return true;

        if(r < 0 || r >= n || c < 0 || c >= m || board[r][c] != word[i]) return false;

        char temp = board[r][c];
        board[r][c] = '0';

        for(auto it : dir){
            int adjr = r + it.first;
            int adjc = c + it.second;
            if(dfs(adjr, adjc, i + 1, word, board, dir)){
                board[r][c] = temp;
                return true;
            }
        }

        board[r][c] = temp;
        return false;
    }

    bool exist(vector<vector<char>>& board, string word) {
        vector<pair<int, int>> dir = {{0,1}, {1,0}, {0,-1}, {-1,0}};

        int n = board.size();
        int m = board[0].size();

        for(int r = 0; r < n; r++){
            for(int c = 0; c < m; c++){
                if(dfs(r, c, 0, word, board, dir)){
                    return true;
                }
            }
        }

        return false;
    }
};

File: 0079-word-search/0079-word-search.py
class Solution(object):
    def __init__(self):
        self.ROWS = 0
        self.COLS = 0
        self.boardCopy = []

    def backtrack(self, row, col, word, index):
        if index >= len(word): 
            return True
        if row < 0 or row == self.ROWS or col < 0 or col == self.COLS or self.boardCopy[row][col] != word[index]:
            return False
        self.boardCopy[row][col] = '#'
        rowOffset = [0,1,0,-1]
        colOffset = [1,0,-1,0]

        for i in range(4):
            result = self.backtrack(row+rowOffset[i], col+colOffset[i], word, index+1)
            if result:
                return True
        self.boardCopy[row][col] = word[index]
        return False 
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        self.boardCopy = board
        self.ROWS = len(board)
        self.COLS = len(board[0])

        for row in range(self.ROWS):
            for col in range(self.COLS):
                if self.backtrack(row, col, word, 0):
                    return True
        return False